src/serialize/value.rs

```diff
@@ -0,0 +1,906 @@
+};
+use scylla_cql::value::{
+    Counter, CqlDate, CqlDecimal, CqlDuration, CqlTime, CqlTimestamp, CqlTimeuuid, CqlValue,
+    CqlVarint,
+};
+use std::net::IpAddr;
+use std::ops::Deref;
+use std::sync::Arc;
+use thiserror::Error;
+use uuid::Uuid;
+
+#[derive(Error, Debug)]
+pub(crate) enum PythonDriverSerializationError {
+    #[error(transparent)]
+    PythonError(#[from] PyErr),
+
+    #[error("Unknown native type")]
+    UnknownNativeType,
+
+    #[error("Unknown collection type")]
+    UnknownCollectionType,
+
+    #[error("Unknown collection type")]
+    UnknownColumnType,
+}
+
+static INT_COLUMNS: &[ColumnType<'static>] = &[
+    ColumnType::Native(NativeType::Int),
+    ColumnType::Native(NativeType::BigInt),
+    ColumnType::Native(NativeType::Date),
+    ColumnType::Native(NativeType::Timestamp),
+    ColumnType::Native(NativeType::Time),
+    ColumnType::Native(NativeType::SmallInt),
+    ColumnType::Native(NativeType::TinyInt),
+];
+
+static FLOAT_COLUMNS: &[ColumnType<'static>] = &[
+    ColumnType::Native(NativeType::Double),
+    ColumnType::Native(NativeType::Float),
+];
+
+static STRING_COLUMNS: &[ColumnType<'static>] = &[
+    ColumnType::Native(NativeType::Ascii),
+    ColumnType::Native(NativeType::Text),
+];
+
+static BOOL_COLUMNS: &[ColumnType<'static>] = &[ColumnType::Native(NativeType::Boolean)];
+
+#[derive(Debug)]
+pub(crate) struct PyAnyWrapper<'py>(pub(crate) Bound<'py, PyAny>);
+
+impl<'py> Deref for PyAnyWrapper<'py> {
+    type Target = Bound<'py, PyAny>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'py> PyAnyWrapper<'py> {
+    fn map_type_to_expected(&self) -> Result<&'static [ColumnType<'static>], SerializationError> {
+        let python_type_name = self
+            .get_type()
+            .name()
+            .map_err(|e| SerializationError::new(PythonDriverSerializationError::PythonError(e)))?;
+
+        let python_type_name = python_type_name
+            .extract::<&str>()
+            .map_err(|e| SerializationError::new(PythonDriverSerializationError::PythonError(e)))?;
+
+        let columns = match python_type_name {
+            "int" => INT_COLUMNS,
+            "float" => FLOAT_COLUMNS,
+            "bool" => BOOL_COLUMNS,
+            "str" => STRING_COLUMNS,
+
+            // Only a subset of Python types is currently supported. More can be added after
+            // deciding on error handling approach.
+            _ => unimplemented!("{:?}", python_type_name),
+        };
+
+        Ok(columns)
+    }
+
+    fn mismatched_type_error(&self, name: &'static str, typ: &ColumnType) -> SerializationError {
+        let expected = match self.map_type_to_expected() {
+            Ok(expected) => expected,
+            Err(e) => return e,
+        };
+        SerializationError::new(mk_typck_err_named(
+            name,
+            typ,
+            BuiltinTypeCheckErrorKind::MismatchedType { expected },
+        ))
```

Comment:
😮 So here we can error out when trying to construct an error. Very Python things...